### DISCUSSION TITLE
Use ESPNow as Transport

### DISCUSSION THREAD

**Discussion by RaphaelDives** [FortySevenEffects/arduino_midi_library#357] [Q&A]
Hey there,
I want to build a wireless transmitter with ESP32 and ESP Now.
ESP Now works like a Serial Connection (already working) with read and write functions. So I think it should not be very hard to implement.
But I don't know where to start. I guess I could just make a loop an send incomming messages through the "SerialNow" object? Or is there a better approach? Like providing the serial object for the MIDI Instance or an ESPNow transport layer or something? Could I just reuse the original SerialMIDI.h and adapt it for the ESP-Now?
Thanks!
Raphael

**Comment by franky47** [comment]
> Could I just reuse the original SerialMIDI.h and adapt it for the ESP-Now?

Yes I think this would be the best approach. Then you'd create an instance of that transport, and pass it to the MIDI Library instance to use. You can see how this is done in the tests, which use a custom dummy transport (to inject input data and observe outputs): https://github.com/FortySevenEffects/arduino_midi_library/blob/master/test/unit-tests/tests/unit-tests_MidiInput.cpp

**Comment by RaphaelDives** [comment]
Thank You very much!

In the end it was easier than I thought - just use the EspNow Serial Class:

ESP_NOW_Serial_Class NowSerial(peer_mac, ESPNOW_WIFI_CHANNEL, ESPNOW_WIFI_IF);
MIDI_CREATE_INSTANCE(ESP_NOW_Serial_Class, NowSerial, MIDI_ESP);

At least I could send some notes over EspNow. (Trying to send some MIDI_USB Input gives me only garbage...)
